src/gradualizer.erl

```diff
@@ -83,23 +80,500 @@ type_check_dir(Dir, Opts) ->
     end.
 
 %% @doc Type check a source or beam file
-%% (Option `print_file' is implicitely true)
--spec type_check_files([file:filename()]) -> ok | nok.
+-spec type_check_files([file:filename()]) -> [{file:filename(), [typechecker:type_error()]}].
 type_check_files(Files) ->
     type_check_files(Files, []).
 
 %% @doc Type check a source or beam
-%% (Option `print_file' is implicitely true)
--spec type_check_files([file:filename()], options()) -> ok | nok.
+-spec type_check_files([file:filename()], typechecker:options()) -> [{file:filename(), [typechecker:type_error()]}].
 type_check_files(Files, Opts) ->
     StopOnFirstError = proplists:get_bool(stop_on_first_error, Opts),
     lists:foldl(
-        fun(File, Res) when Res =:= ok;
+        fun(File, Errors) when Errors =:= [];
                             not StopOnFirstError ->
-                case type_check_file(File, [print_file|Opts]) of
-                    ok -> Res;
-                    nok -> nok
+                case type_check_file(File, Opts) of
+                    [] -> Errors;
+                    NewErrors -> [{File, NewErrors} | Errors]
                 end;
-            (_, nok) ->
-                nok
-        end, ok, Files).
+            (_, Errors) ->
+                Errors
+        end, [], Files).
+
+-spec format_error(typechecker:type_error()) -> string().
```

Comment:
`io_lib:format` doesn't return a string() but rather an iolist(), a nested list of strings.